quickwit/quickwit-rest-client/src/rest_client.rs

```diff
@@ -0,0 +1,698 @@
+        if let Some(bytes) = body {
+            request_builder = request_builder.body(bytes);
+        };
+        if let Some(qs) = query_string {
+            request_builder = request_builder.query(qs);
+        }
+        let response = request_builder.send().await?;
+
+        Ok(ApiResponse::new(response))
+    }
+}
+
+/// Root client for top level APIs.
+pub struct QuickwitClient {
+    transport: Transport,
+}
+
+impl QuickwitClient {
+    pub fn new(transport: Transport) -> Self {
+        Self { transport }
+    }
+
+    pub async fn search(
+        &self,
+        index_id: &str,
+        search_query: SearchRequestQueryString,
+    ) -> Result<SearchResponseRest, Error> {
+        let path = format!("{index_id}/search");
+        let bytes = serde_json::to_string(&search_query)
+            .unwrap()
+            .as_bytes()
+            .to_vec();
+        let body = Bytes::from(bytes);
+        let response = self
+            .transport
+            .send::<()>(Method::POST, &path, HeaderMap::new(), None, Some(body))
+            .await?;
+        let search_response = response.deserialize().await?;
+        Ok(search_response)
+    }
+
+    pub fn indexes(&self) -> IndexClient {
+        IndexClient::new(&self.transport)
+    }
+
+    pub fn splits<'a, 'b: 'a>(&'a self, index_id: &'b str) -> SplitClient {
+        SplitClient::new(&self.transport, index_id)
+    }
+
+    pub fn sources<'a, 'b: 'a>(&'a self, index_id: &'b str) -> SourceClient {
+        SourceClient::new(&self.transport, index_id)
+    }
+
+    pub async fn ingest(&self, index_id: &str, ingest_source: IngestSource) -> Result<(), Error> {
+        let ingest_path = format!("{index_id}/ingest");
+        let reader: Box<dyn AsyncRead + Send + Sync + Unpin> = match ingest_source {
+            IngestSource::File(filepath) => {
+                let file = File::open(&filepath).await?;
+                Box::new(file)
+            }
+            IngestSource::Stdin => Box::new(tokio::io::stdin()),
+        };
+        let mut buf_reader = BufReader::new(reader);
+
+        let start = Instant::now();
+        let mut num_bytes_sent = 0;
+
+        loop {
+            let mut buffer = Vec::with_capacity(4 * 1024 * 1024);
+            for _ in 0..5_000 {
+                if buf_reader.read_until(b'\n', &mut buffer).await? == 0 {
+                    break;
+                }
+            }
+
+            if buffer.is_empty() {
+                break;
+            }
+            let body = Bytes::from(buffer);
+
+            loop {
+                let response = self
+                    .transport
+                    .send::<()>(
+                        Method::POST,
+                        &ingest_path,
+                        HeaderMap::new(),
+                        None,
+                        Some(body.clone()),
+                    )
+                    .await?;
+
+                if response.status_code() == 429 {
```

Comment:
let's use the defined status code constant `hyper::StatusCode::TOO_MANY_REQUESTS`